### PR TITLE
cmake: Do a debug build by default

### PR DESCRIPTION
**Pull request by badone**
Signed-off-by: Brad Hubbard <bhubbard@redhat.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`

</details>

### CODE REVIEW

**Comment by badone** [comment]
Related to https://github.com/ceph/ceph/pull/30250

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -74,7 +74,7 @@ endif(CMAKE_CXX_COMPILER_ID STREQUAL GNU)
 set(CMAKE_CXX_FLAGS "${CMAKE_C_FLAGS} ${CMAKE_CXX_FLAGS}")
 
 if(NOT CMAKE_BUILD_TYPE)
-  set(CMAKE_BUILD_TYPE "RelWithDebInfo" CACHE STRING "Default BUILD_TYPE is RelWithDebInfo, other options are: Debug, Release, and MinSizeRel." FORCE)
+  set(CMAKE_BUILD_TYPE "Debug" CACHE STRING "Default BUILD_TYPE is Debug, other options are: RelWithDebInfo, Release, and MinSizeRel." FORCE)
```
Comment (tchaikov) [inline src/CMakeLists.txt]: this changes how the rpmbuild builds ceph packages. before this change, it builds with `CMAKE_BUILD_TYPE=RelWithDebInfo`, after this change it builds with `CMAKE_BUILD_TYPE=Debug`, which is not expected.

unlike debian/rules where we use the default parametersdh helper specified by `dh_auto_configure`, which passes `-DCMAKE_BUILD_TYPE=None` explicitly, `ceph.spec.in` does not pass this setting to cmake. i also checked "cmake-rpm-macros" package for a reference, neither does set `CMAKE_BUILD_TYPE`. see https://github.com/open-io/rpm-specfiles/blob/master/cmake/macros.cmake for a (probably outdated) copy of the macros defined by it. BTW, we are not using any of these macros, as this package is not available in RHEL / CentOS 7.

so i think a better fix is
```cmake

if(NOT CMAKE_BUILD_TYPE)
  if(EXISTS "${CMAKE_SOURCE_DIR}/.git")
    set(default_build_type "Debug")
  else()
    set(default_build_type "RelWithDebInfo")
  endif()
  set(CMAKE_BUILD_TYPE "${default_build_type}" CACHE
      STRING "Default BUILD_TYPE is Debug, other options are: RelWithDebInfo, Release, and MinSizeRel." FORCE)
endif()
```

Comment (badone) [reply src/CMakeLists.txt]: Yes, this looks much better. Thanks a lot @tchaikov

File: src/CMakeLists.txt [hunk 2]
```diff
@@ -74,7 +74,13 @@ endif(CMAKE_CXX_COMPILER_ID STREQUAL GNU)
...
 
 if(NOT CMAKE_BUILD_TYPE)
-  set(CMAKE_BUILD_TYPE "RelWithDebInfo" CACHE STRING "Default BUILD_TYPE is RelWithDebInfo, other options are: Debug, Release, and MinSizeRel." FORCE)
+  if(EXISTS "${CMAKE_SOURCE_DIR}/.git")
+    set(default_build_type "Debug")
+  else()
+    set(default_build_type "RelWithDebInfo")
+  endif()
+  set(CMAKE_BUILD_TYPE "${default_build_type}" CACHE
+      STRING "Default BUILD_TYPE is Debug, other options are: RelWithDebInfo, Release, and MinSizeRel." FORCE)
```
Comment (tchaikov) [inline src/CMakeLists.txt]: i was thinking if we need to tweak this string. but i think, in general, this message is useful only to cmake GUI users, who are very likely to be developers working on a git repo. so "Debug" and "RelWithDebInfo, Release, and MinSizeRel" should be good enough.

**Comment by tchaikov** [comment]
https://shaman.ceph.com/builds/ceph/wip-kefu-testing-2019-10-09-1243/0dca79308d83afd02b99aa1aa8839d156215298c/

**Pull request merged**